astrocut/make_cube.py

```diff
@@ -346,3 +349,496 @@ def make_cube(self, file_list, cube_file="img-cube.fits", sector=None, max_memor
+        self.block_size = None  # Number of rows
+        self.num_blocks = None
+        self.cube_shape = None
+        
+        self.time_keyword = 'STARTTJD'  # Start time in TJD. TICA-specific.
+        self.last_file_keywords = ['ENDTJD']  # Stop time in TJD. TICA-specific (assumed to be in extension 0)                  
+        self.image_header_keywords = ['CAMNUM', 'CCDNUM']  # Camera number and CCD number 
+        self.template_requirements = {'NAXIS': 2}  # Using NAXIS instead of WCSAXES. JENNY: Verify this is OK.
+                                                   
+        self.file_list = None
+        self.template_file = None
+        
+        self.primary_header = None
+        self.info_table = None
+        self.cube_file = None
+
+        # Used when updating an existing cube 
+        self.old_cols = None
+        self.update = False
+        self.cube_append = None
+
+    def _configure_cube(self, file_list, **extra_keywords):
+        """ Run through all the files and set up the  basic parameters for the cube.
+        Set up the cube primary header.
+        """
+
+        file_list = np.array(file_list)
+        image_shape = None
+        start_times = np.zeros(len(file_list))
+        for i, ffi in enumerate(file_list):
+            ffi_data = fits.open(ffi, mode='denywrite', memmap=True)
+            
+            start_times[i] = ffi_data[0].header.get('STARTTJD')
+
+            if image_shape is None:  # Only need to fill this once
+                image_shape = ffi_data[0].data.shape
+            
+            if self.template_file is None:  # Only check this if we don't already have it
+
+                is_template = True
+                for key, value in self.template_requirements.items():
+                    if ffi_data[0].header.get(key) != value:  # Checking for a good image header
+                        is_template = False
+                        
+                if is_template:
+                    self.template_file = ffi
+                    
+            ffi_data.close()
+
+        self.file_list = file_list[np.argsort(start_times)]
+
+        # Working out the block size and number of blocks needed for writing the cube
+        # without using too much memory
+        slice_size = image_shape[1] * len(self.file_list) * 2 * 4  # in bytes (float32)
+        max_block_size = int((self.max_memory * 1e9)//slice_size)
+        
+        self.num_blocks = int(image_shape[0]/max_block_size + 1)
+        self.block_size = int(image_shape[0]/self.num_blocks + 1)
+        
+        # Determining cube shape: 
+        # If it's a new cube, the shape is (nRows, nCols, nImages, 2)
+        if not self.update:
+            self.cube_shape = (image_shape[0], image_shape[1], len(self.file_list), 2)
+            
+        # Else, if it's an update to an existing cube, the shape is (nRows, nCols, nImages + nNewImages, 2)
+        else: 
+            self.cube_shape = self.cube_append.shape
```

Comment:
Just wondering why this is different compared to the default `CubeFactory`

In the original factory, it seems like cube_shape is always `(image_shape[0], image_shape[1], len(self.file_list), 2)`

do those no receive updates?

Reply:
This new logic + the everything under the TICA-specific function called `update_cube` is so that the TESS team can add new FFIs onto existing TICA cubes as they get delivered. SPOC deliveries will continue to be only once per month for a single sector, whereas TICA deliveries are 2x a month and will increase to 4x a month starting in September. 

We will make a fresh TICA cube at the beginning of the month so we can get data out faster to users, but we will need to update these cubes as new data comes in throughout the month for a given sector, so that's what this logic is for, and is only specific to TICA.